src/model_theory/basic.lean

```diff
@@ -313,5 +331,242 @@ lemma comp_assoc (f : M ≃[L] N) (g : N ≃[L] P) (h : P ≃[L] Q) :
+
+/-- Copy a substructure replacing `carrier` with a set that is equal to it. -/
+protected def copy (S : L.substructure M) (s : set M) (hs : s = S) : L.substructure M :=
+{ carrier := s,
+  fun_mem := λ n f, hs.symm ▸ (S.fun_mem f) }
+
+variable {S : L.substructure M}
+
+@[simp] lemma coe_copy {s : set M} (hs : s = S) :
+  (S.copy s hs : set M) = s := rfl
+
+lemma copy_eq {s : set M} (hs : s = S) : S.copy s hs = S :=
+set_like.coe_injective hs
+
+lemma const_mem {c : L.const} : ↑c ∈ S :=
+mem_carrier.2 (S.fun_mem c _ fin.elim0)
+
+/-- The substructure `M` of the structure `M`. -/
+instance : has_top (L.substructure M) :=
+⟨{ carrier := set.univ,
+   fun_mem := λ n f x h, set.mem_univ _ }⟩
+
+instance : inhabited (L.substructure M) := ⟨⊤⟩
+
+@[simp] lemma mem_top (x : M) : x ∈ (⊤ : L.substructure M) := set.mem_univ x
+
+@[simp] lemma coe_top : ((⊤ : L.substructure M) : set M) = set.univ := rfl
+
+/-- The inf of two substructures is their intersection. -/
+instance : has_inf (L.substructure M) :=
+⟨λ S₁ S₂,
+  { carrier := S₁ ∩ S₂,
+    fun_mem := λ n f, closed_under_inf (S₁.fun_mem f) (S₂.fun_mem f) }⟩
+
+@[simp]
+lemma coe_inf (p p' : L.substructure M) : ((p ⊓ p' : L.substructure M) : set M) = p ∩ p' := rfl
+
+@[simp]
+lemma mem_inf {p p' : L.substructure M} {x : M} : x ∈ p ⊓ p' ↔ x ∈ p ∧ x ∈ p' := iff.rfl
+
+instance : has_Inf (L.substructure M) :=
+⟨λ s, { carrier := ⋂ t ∈ s, ↑t,
+        fun_mem := λ n f, closed_under_Inf begin
+          rintro _ ⟨t, rfl⟩,
+          simp only,
+          by_cases h : t ∈ s,
+          { rw [set.Inter_pos h],
+            exact t.fun_mem f },
+          { rw [set.Inter_neg h],
+            exact λ _ _, set.mem_univ _ }
+        end }⟩
+
+@[simp, norm_cast]
+lemma coe_Inf (S : set (L.substructure M)) :
+  ((Inf S : L.substructure M) : set M) = ⋂ s ∈ S, ↑s := rfl
+
+lemma mem_Inf {S : set (L.substructure M)} {x : M} : x ∈ Inf S ↔ ∀ p ∈ S, x ∈ p :=
+  set.mem_bInter_iff
+
+lemma mem_infi {ι : Sort*} {S : ι → L.substructure M} {x : M} : (x ∈ ⨅ i, S i) ↔ ∀ i, x ∈ S i :=
+by simp only [infi, mem_Inf, set.forall_range_iff]
+
+@[simp, norm_cast]
+lemma coe_infi {ι : Sort*} {S : ι → L.substructure M} : (↑(⨅ i, S i) : set M) = ⋂ i, S i :=
+by simp only [infi, coe_Inf, set.bInter_range]
+
+/-- Substructures of a structure form a complete lattice. -/
+instance : complete_lattice (L.substructure M) :=
+{ le           := (≤),
+  lt           := (<),
+  top          := (⊤),
+  le_top       := λ S x hx, mem_top x,
+  inf          := (⊓),
+  Inf          := has_Inf.Inf,
+  le_inf       := λ a b c ha hb x hx, ⟨ha hx, hb hx⟩,
+  inf_le_left  := λ a b x, and.left,
+  inf_le_right := λ a b x, and.right,
+  .. complete_lattice_of_Inf (L.substructure M) $ λ s,
+    is_glb.of_image (λ S T,
+      show (S : set M) ≤ T ↔ S ≤ T, from set_like.coe_subset_coe) is_glb_binfi }
+
+variable (L)
+
+/-- The `L.substructure` generated by a set. -/
+def closure (s : set M) : L.substructure M := Inf {S | s ⊆ S}
```

Comment:
It might be worth the effort to instead define `closure` as a `lower_adjoint (coe : L.substructure M → set M)` and get a lot of the following lemmas (and quite a few additional ones!) for free. However, `lower_adjoint` doesn't seem to be used yet in mathlib, so it's fine if you want to leave that to someone else.

Reply:
Done!

There's probably a better way to connect `lower_adjoint` to `gi`, but it does require an additional assumption (that the upper adjoint has the property that `u x <= u y` implies `x <= y`). I think this could at least be done at the level of `set_like`, but perhaps that's for a different PR.